build(broccoli): add tree-stabilizer plugin to deal with unstable trees

IgorMinar opened the pull request:
Previously we assumed that all input and ouput paths for broccoli trees are immutable, that turned out to be
incorrect.

By adding a tree stabilizer plugin in front of each diffing plugin, we ensure that the input trees
are stable. The stabilization is done via symlinks which is super cheap on platforms that support
symlinks. On Windows we currently copy the whole input directory, which is far from ideal. We should
investagate if using move operation on Windows is ok in the future to improve performance.

Closes #2051

IgorMinar committed:
build(broccoli): add tree-stabilizer plugin to deal with unstable trees

Previously we assumed that all input and ouput paths for broccoli trees are immutable, that turned out to be
incorrect.

By adding a tree stabilizer plugin in front of each diffing plugin, we ensure that the input trees
are stable. The stabilization is done via symlinks which is super cheap on platforms that support
symlinks. On Windows we currently copy the whole input directory, which is far from ideal. We should
investagate if using move operation on Windows is ok in the future to improve performance.

Closes #2051

caitp commented on tools/broccoli/broccoli-tree-stabilizer.ts:
```diff
@@ -0,0 +1,51 @@
+    if (this.previousInputPath ===  this.inputPath) return;
+
+    this.previousInputPath =  this.inputPath;
+    fs.rmdirSync(this.outputPath);
```
rmdirSync will fail if non-empty, which it probably is. use rimraf, which we already depend on

caitp replied:
Also, if outputPath is a simlink of inputPath, isn't that going to remove everything in the inputPath?

joliss replied:
Calling rimraf (or fs-extra, which calls rimraf) on a symlink will only remove the symlink, not descend into the target directory.

IgorMinar replied:
@caitp unless there is a bug in the system, we can't get into a situation where the underlying directory doesn't exist. output path is created by broccoli builder for each rebuild unless it already exists.

caitp replied:
That doesn't address what my comment is saying. "rmdirSync" will fail (read: throw) if the directory is non-empty. We aren't talking about whether it exists or not, we're talking about it being non-empty, which it probably is. This most likely works correctly on unix systems where symlinkOrCopy is not copying, but on windows machines it's likely to break

caitp commented on tools/broccoli/broccoli-tree-stabilizer.ts:
```diff
@@ -0,0 +1,51 @@
+
+    // TODO: investagate if we can use rename the directory instead to improve performance on
+    // Windows
+    symlinkOrCopy.sync(this.inputPath, this.outputPath);
```
This extra tree is just moving the workaround that's already happening elsewhere, which sort of seems like a non-solution, to be honest. I think we're better off talking to @joliss about making sure the contracts of the inputPath(s) and outputPath/cachePath properties make sense and are stable across versions

IgorMinar replied:
to be clear, this change is just a workaround for incorrect behavior in some broccoli plugins.

my goal is to replace all of the significant plugins in our build chain with diffing plugins, and for those we can skip this operation.

as @joliss confirmed, eventually this should not be a problem for any plugins as they'll get rewritten to work differently.

caitp commented on tools/broccoli/broccoli-tree-stabilizer.ts:
```diff
@@ -0,0 +1,51 @@
+ *
+ * For example all `DiffingBroccoliPlugin`s expect their inputPath to be stable.
+ *
+ * Be inserting this plugin into the tree after any misbehaving plugin, we can stabilize the
```
typo: "By"

IgorMinar replied:
fixed

caitp commented:
For the record, I think we see better perf on win32 by avoiding the deep copy and just dealing with the property changing. Deep copy could be quite expensive in some cases

caitp commented:
Another option is to just re-initialize the wrapped plugin if the inputPath(s) change, so you still get to keep your happy lifecycle management, without having to do unnecessary copying

IgorMinar committed:
build(broccoli): TreeDiffer should log associated plugin name rather than inputPath

Usually we don't care what we are diffing, but why we are diffing it. With this change we see what is causing build slowdown
due to diffing.

IgorMinar commented:
@caitp we can't re-init the plugin as that would kill the incremental build.

IgorMinar committed:
build(broccoli): update broccoli.d.ts docs

IgorMinar committed:
build(gulp): remove obsolete watch.js.dev build task

IgorMinar committed:
build(broccoli): improve error messaging from TreeDiffer

IgorMinar committed:
style(build): reorder imports

IgorMinar committed:
build(broccoli): reorder tree-differ specs
